### ISSUE TITLE
Allow custom bins for numerical fields

### ISSUE THREAD

**Issue by vitorbaptista** [metabase/metabase#12847]
**Is your feature request related to a problem? Please describe.**

The auto-generated numerical column bins don't always fit business rules.

**Describe the solution you'd like**

I'd like to be able to configure custom bins. They could be completely customizable e.g. by giving a list of numbers to be the edges of the bins. Any number lower than the lowest edge is grouped together, likewise for numbers higher than the highest edge.

Alternatively, we could have just:

* Minimum value
* Maximum value
* Number of bins

So the buckets are "below minimum value", "above maximum value", and the range between min and max are split in X equal-sized bins. This is less flexible, as all bins will have the same width, but it's good enough for many cases and the UI is simpler (only 3 numbers, instead of N).

I based these parameters on the [WIDTH_BUCKET](https://docs.snowflake.com/en/sql-reference/functions/width_bucket.html) Snowflake function.

**Describe alternatives you've considered**

Right now, to create custom sized bins, I can:

* Make a SQL question - Bad because it requires SQL knowledge and I miss all the goodies of the query generator
* Add a new column with the bin groups - Bad because it requires adding a new column and the bins are pre-defined.

**How important is this feature to you?**

I need these custom bins to create some reports. As Metabase doesn't support it, and my target audience doesn't know SQL, I'm manually generating columns like `bin_width_5_max_15` with the bin mappings. This isn't ideal.

**Additional context**

This is how Tableau handles it: https://help.tableau.com/current/pro/desktop/en-us/calculations_bins.htm.

**Comment by sbelak** [comment]
You can use custom expression with `case` https://www.metabase.com/docs/latest/users-guide/expressions.html to define custom bins.

**Comment by vitorbaptista** [comment]
@sbelak Thanks for the quick response. I added it to the alternatives list on the first message.

**Comment by flamber** [comment]
This is likely a duplicate of #9208, which can be difficult to overview, but there are some good references.
Number of bins can be adjusted - default is `8`: https://github.com/metabase/metabase/blob/master/docs/operations-guide/environment-variables.md#mb_breakout_bins_num

**Comment by vitorbaptista** [comment]
@flamber You're right, this seems like a duplicate of #9208. That issue talks about maps, which muddles the message a bit, but the point is to allow defining the bins using:

* Minimum value
* Maximum value
* Bin width

If you think that's better, feel free to close this issue and let's discuss on #9208.

If I find the time to work on it (and learn Clojure), is this an issue you'd like fixed? The main challenge for me would be the UI.

**Comment by flamber** [comment]
@vitorbaptista I kept this open, since it's very well described :+1: 
It is definitely something that has been requested, but it involves several areas, like you say, so I don't think it's something that can be solved by one person alone - and we want the UI to stay simple, which is probably the biggest challenge.
I would recommend other, small issues to start with, so you get the hang of the code base and doesn't turn into a massive project.

**Comment by EBoisseauSierra** [comment]
I think we could generalise this a notch more by considering the following. There are 4 parameters to create an histogram:
* min value,
* max value,
* bin “width”,
* count of bins.

… yet there are only 3 degrees of freedom (i.e. the fourth value is dependent from the 3 other ones).

Min and max values can be defaulted (as it currently is) to variable's min and max value — modifying this can be currently be achieved via filters (to artificially change the variable's min and max). Being able to do this at the aggregation step would be nice, but might create an overly cumbersome UI.

So what would be really nice is to have the choice to:
* either **manually select the number of bins** — rather than choosing from a fixed list as it currently is the case — (and this defines the bin width as `(max − min) ÷ n`),
* or **manually set the bin width** (so that we have a number `n` of bin such as `min + n × width ≥ max` — using the lowest `n` possible value).

**Comment by jag2136** [comment]
Quick example of why a more customizable bin scheme (to fit specific business cases) would be helpful:

<img width="1240" alt="Screen Shot 2021-07-25 at 9 27 37 PM" src="https://user-images.githubusercontent.com/7367690/126933073-e0c874b4-1fe1-4a7e-bd97-72e5b0ed1fc8.png">

Profile scores range between 0 and 100 (it's a %), yet the bins generated (both 10 and auto) include values over 100. The highest profile scores ARE 100, so they fit into 100-110, but 100-110 makes no sense as a bin and makes the chart less useful. I even set a "Profile score <= 100" filter in case that would affect the binning, but no luck.

Would love to set my own arbitrary max bin.

**Comment by codingjoe** [comment]
Just to throw this into the mix here, it would also already be an improvement, if you'd be able to switch between linear, exponential or logarithmic bins. That would probably cover numerous cases, since natural distribution of number usually comes in one of those forms. With the added bonus, that this would be fairly simple to implement.

**Comment by egluhbegovic** [comment]
> 

@sbelak Your answer solves the problem elegantly.
Thank you!
